src/io/flutter/run/test/TestForm.java

```diff
@@ -97,11 +105,41 @@ private Scope getScope() {
     return (Scope)scope.getSelectedItem();
   }
 
-  private void render() {
-    testDirLabel.setVisible(getScope() == Scope.DIRECTORY);
-    testDir.setVisible(getScope() == Scope.DIRECTORY);
+  /**
+   * When switching between file and directory scope, update the next field to
+   * a suitable default.
+   */
+  private void updateFields(Scope next) {
+    if (next == Scope.DIRECTORY && displayedScope == Scope.FILE) {
+      final String sep = String.valueOf(File.separatorChar);
+
+      final String path = testFile.getText();
+      if (path.contains(sep) && path.endsWith(".dart")) {
+        // Remove the last part of the path to get a directory.
+        testDir.setText(path.substring(0, path.lastIndexOf(sep) + 1));
+      } else if (testDir.getText().isEmpty()) {
+        // Keep the same path; better than starting blank.
+        testDir.setText(path);
+      }
+
+    } else if (next == Scope.FILE && displayedScope == Scope.DIRECTORY) {
+      if (testFile.getText().isEmpty()) {
+        testFile.setText(testDir.getText());
+      }
+    }
+  }
+
+  /**
+   * Show and hide fields as appropriate for the next scope.
+   */
+  private void render(Scope next) {
+
+    testDirLabel.setVisible(next == Scope.DIRECTORY);
```

Comment:
Probably 6 of one and 1/2 dozen of another, but you could also one path field component, and toggle it's behavior when you hit the browse button (instead of showing and hiding sets of fields)..